### PR TITLE
add pci bridge to guests

### PR DESCRIPTION
**Pull request by tbaev**
Fix for broken virt-manager test. A new needle is required to clean up the rest of the soft failures but currently we have to resolve [poo#106101](https://progress.opensuse.org/issues/106101) first.

- Related ticket: https://progress.opensuse.org/issues/101536
- Needles: N/A
- Verification run: [sle-15-SP1 XEN](http://openqa.qam.suse.cz/tests/36794#dependencies)
[sle-15-SP1 KVM](http://openqa.qam.suse.cz/tests/36896#dependencies)
[sle-15-SP2 KVM](http://openqa.qam.suse.cz/tests/36810#dependencies)
[sle-15-SP3 KVM](http://openqa.qam.suse.cz/tests/36820#dependencies)
[sle-12-SP4 KVM](http://openqa.qam.suse.cz/tests/36830#dependencies)
[sle-15 KVM](http://openqa.qam.suse.cz/tests/36910#dependencies)

### CODE REVIEW

File: data/virtualization/add_pci_bridge.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+machine="$1"
+xml="$2"
+
+if [[ $machine == "q35" ]]; then
+	# add 10 pcie-root-port devices for hotplugging
+	for n in {1..10}
+		do
+		sed -i "/.*<\/devices>/i<controller type='pci' model='pcie-root-port'/>" $xml
+	        done
```
Comment (alice-suse) [inline data/virtualization/add_pci_bridge.sh:15]: Please pay attention to indent for above 3 lines.

File: data/virtualization/add_pci_bridge.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+# Add pci devices to guest .xml to enable hotplugging. This script can be expanded to add custom number
+# of devices depending on specific tests cases.
+# Maintainer: Teodor Baev <teodor.baev@suse.com>
+
+machine="$1"
+xml="$2"
+
+if [[ $machine == "q35" ]]; then
+	# add 10 pcie-root-port devices for hotplugging
+	for n in {1..10}
```
Comment (alice-suse) [inline data/virtualization/add_pci_bridge.sh:12]: I am curious about the minimum required pcie-root-port for our hotplugging test and how you decide to add 10 ports?

**Comment by alice-suse** [comment]
@tbaev Thanks for the PR. I checked the validation jobs -- we can finally see them green again , although as you mentioned some needs more needle work. Good job!

File: tests/virtualization/universal/waitfor_guests.pm [hunk 3]
```diff
@@ -40,10 +40,10 @@ sub add_pci_bridge {
...
         return if (script_run("cat $xml | grep 'controller' | grep 'pci-bridge'") == 0);
 
-        my $regex = '<controller type=.pci. .*model=.pci-root..*>';
-        my $addbefore = '<controller type="pci" model="pci-bridge"/>';
-        # the add_before.sh script inserts a given line before the line matching the given regex
-        assert_script_run("virsh dumpxml $guest | /root/add_before.sh '$regex' '$addbefore' > $xml");
+        my $machine = 'i440fx';
+        # add_pci_bridge.sh script inserts 1 pci-bridge for i400fx machine.
+        assert_script_run("virsh dumpxml $guest > $xml");
+        assert_script_run("/root/add_pci_bridge.sh '$machine' '$xml'");
```
Comment (pdostal) [inline tests/virtualization/universal/waitfor_guests.pm:46]: Why is it important that this functionality is in bash? Wouldn't be native perl subroutine better?

Comment (pdostal) [reply tests/virtualization/universal/waitfor_guests.pm:46]: Plus, there is already nice `add_pci_bridge()` subroutine :smile:

**Comment by tbaev** [comment]
`add_pci_bridge()` subroutine is the one that includes the bash script (before and after my changes), or you are talking about another subroutine for adding bridges?

It was done in bash previously from the `add_before.sh` script, that I started modifying and ending by just recreating it. 

It can be done in perl but it will take a lot of time for openQA to type the sed command multiple times for all the guests. This is the only meaningful benefit I can see for using a bash script.

**Comment by pdostal** [comment]
It will not take more than 1 second with the serial console which we use.
The benefits are readability and observability.

**Comment by tbaev** [comment]
Oh I didn't know that. I thought we are limited at the "type speed" that I was observing when using VNC at `boot_from_pxe`, for example.

I will correct that and re-write it only in Perl, it will be better. :)

**Comment by pdostal** [comment]
Then booting and the installation is running via VNC but then you can see that there is no needles, just console output :)

http://openqa.qam.suse.cz/tests/36906

File: tests/virtualization/universal/waitfor_guests.pm [hunk 4]
```diff
@@ -60,10 +59,12 @@ sub add_pci_bridge {
         # Skip if a pcie-to-pci-bridge is already present
         return if (script_run("cat $xml | grep 'controller' | grep 'pcie-to-pci-bridge'") == 0);
 
-        my $regex = '<controller type=.pci. .*model=.pcie-root..*>';
-        my $addbefore = '<controller type="pci" model="pcie-root-port"/><controller type="pci" model="pcie-to-pci-bridge"/>';
-        # the add_before.sh script inserts a given line before the line matching the given regex
-        assert_script_run("virsh dumpxml $guest | /root/add_before.sh '$regex' '$addbefore' > $xml");
+        # Add 10 pcie-root-port devices and a pcie-to-pci brdige
+        assert_script_run("virsh dumpxml $guest > $xml");
+        for (1 .. 10) {
```
Comment (alice-suse) [inline tests/virtualization/universal/waitfor_guests.pm:64]: @tbaev  I am curious about the minimum required pcie-root-port for our hotplugging test and how you decide to add 10 ports?

**Comment by tbaev** [comment]
@alice-suse I think 3-4 pcie-root-ports are enough. I can run some test to find out the exact number?. I just pick 10 to have a big headroom for possible future tests.

**Comment by alice-suse** [comment]
> @alice-suse I think 3-4 pcie-root-ports are enough. I can run some test to find out the exact number?. I just pick 10 to have a big headroom for possible future tests.

No, it is not needed, and your decision is reasonable.

**Pull request merged**